### PR TITLE
ninja/meson

### PR DESCRIPTION
**Pull request by dhewg**
This adds ninja/meson as host as well as target packages.

ninja.mk/meson.mk are provided for packages to use these tools, see the comments there how to use it.

libmpdclient/mpc are switched over from the gross autotools hack to meson. No further changes, but this should unblock updating the packages as well as sound/mpd.

Tested on boneblack and cubieboard2 (where one meson cross file is created per target in the host staging dir).

not tested:
- ~~c++ lib != libstdcpp~~ done, ninja can only be built against libstdcpp
- ~~meson builds where native tools are built and executed during the build~~ most likely done, added support for host packages, which uses the same config

not done:
- ~~ccache~~ done, works

cc:
@br101 @ja-pa (earlier attempts openwrt/openwrt#2147 openwrt/openwrt#2312
@ynezz (handling the earlier attempts)
@diizzyy (using ninja with cmake)
@tstenner @lucize @flyn-org @neilwhelchel (building packages using meson)
@thess (libmpdclient/mpc/mpd maintainer)

### CODE REVIEW

**Comment by dhewg** [comment]
I'm unsure where the build failure comes from:
```
Package ninja is missing dependencies for the following libraries:
libstdc++.so.6
```

Seems to work on my end?
```
define Package/ninja
  ...
  DEPENDS:=$(CXX_DEPENDS)
endef

```

```
$ opkg info ninja
Package: ninja
Version: 1.9.0-1
Depends: libc, libstdcpp6
Status: install user installed
Architecture: arm_cortex-a7_neon-vfpv4
Installed-Time: 1569016828
```

**Comment by hnyman** [comment]
cc @thess who is the maintainer of the packages you propose to switch for meson build system introduced by this PR.

File: devel/meson/Makefile [hunk 1]
```diff
@@ -0,0 +1,76 @@
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=meson
+PKG_VERSION:=0.51.0
```
Comment (ja-pa) [inline devel/meson/Makefile]: Last version is 0.51.2

**Comment by ja-pa** [comment]
I feel that a better way would be to introduce switch to meson in separate PR.

**Comment by neheb** [comment]
I doubt uClibc++ can be used here. It's a C++98 library. C++11 or later is probably what's used.

**Comment by thess** [comment]
@ja-pa - agree. Adding/updating packages to use Ninja/Meson needs to be done separately.

@hnyman - Will be investigating mpd,etc. this week.

Any reason why Ninja/Meson devel tools shouldn't be either:
 a) Host build system requirement and/or
 b) Part of toolchain/SDK ala _cmake_.

**Comment by hnyman** [comment]
> Any reason why Ninja/Meson devel tools shouldn't be either:
> a) Host build system requirement and/or
> b) Part of toolchain/SDK ala cmake.

scons was just recently removed from host tools and moved to packages feed by 7087efd72 and https://github.com/openwrt/openwrt/commit/60eb0e8da2aa5df7cbbea657d6a809b29e818ad8 authored by @ynezz , so I think that devs have differing opinions on the approach regarding additional build tools ...

File: devel/meson/Makefile (hunk 1 above)
Comment (dhewg) [reply devel/meson/Makefile]: Oh, right. Updated, thanks!

**Comment by dhewg** [comment]
updated:
- update meson to 0.51.2
- use meson's "native file" feature for host builds to shorten cmdlines
- move libmpdclient/mpc commits to #10052

**Comment by neheb** [comment]
I'm fairly confused on the issue of _HOST vs. _HOSTPKG. On IRC I've been told something equivalent to "_HOST should be used by packages in the tools/ directory".

If we look at this tree, including this PR, we get

```
git grep STAGING_DIR_HOST\) | wc -l
15
```

while _HOSTPKG has this:

```
git grep STAGING_DIR_HOSTPKG\) | wc -l
110
```

**Comment by dhewg** [comment]
I got the a similar response on irc.

updated:
- switched to STAGING_DIR_HOSTPKG

**Comment by neheb** [comment]
ping @jow-

**Comment by dhewg** [comment]
updated:
- fix including ninja.mk from meson.mk if the package subdir depth != 2 (use MAKEFILE_LIST instead of CURDIR)
- add support for building host packages using meson (proper uninstall from hostpkg requires this patch https://patchwork.ozlabs.org/patch/1168833/ though, but it's doesn't break without it)

host builds were tested with [0001-XXX-meson-host-test.patch.txt](https://github.com/openwrt/packages/files/3665418/0001-XXX-meson-host-test.patch.txt)

@thess did you give this a try yet?

**Comment by dhewg** [comment]
updated:
* create native/cross files for each package in [Host|Build]/Configure, as they may need to modify compiler flags (as e.g. nls.mk does)

**Comment by dhewg** [comment]
Here's a tree that includes #10043 #10113 and above mentioned host build test patch.
https://github.com/dhewg/packages/tree/meson

**Comment by dhewg** [comment]
Oh... well, if noone steps forward I guess I can take over mpd related packages for a while.

How to we proceed with this MR? I'm happy with it, the foundation is there and it works nicely for me for #10052. For other packages we may need more features, but that can be dealt with as issues pop up.

Another candidate to test drive this glib2, since v2.59.0 is the last version supporting autotools. @tripolar did you try this yet to build newer versions using meson?

**Comment by neheb** [comment]
AFAIK the odd glib releases are unstable ones.

**Comment by dhewg** [comment]
updated:
* meson v0.52.0

**Comment by flyn-org** [comment]
I can update libsoup, grilo, and grilo-plugins once there exists an example of "porting" a package to use the project's Meson-based build.

@neheb, glib2 2.59 is indeed a development version, but the subsequent 2.60 and 2.62 series also use Meson.

**Comment by dhewg** [comment]
@flyn-org have a look at `devel/meson/meson.mk` from this MR: .https://github.com/openwrt/packages/pull/10043/commits/84d7a513937f46c1b562ee483c5c67e1b2822687#diff-398d125edadb8ea15ae979ed80f5258e
There're also a few packages I ported in this tree https://github.com/dhewg/packages/commits/meson

File: devel/meson/Makefile [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://codeload.github.com/mesonbuild/meson/tar.gz/$(PKG_VERSION)?
+PKG_HASH:=0f426ed1362c38bcc5b9027ec6aec3445d6db88e8d7249ed992e9af88a42d0e0
+
+PKG_MAINTAINER:=Andre Heider <a.heider@gmail.com>
+PKG_LICENSE:=Apache-2.0
+PKG_LICENSE_FILES:=COPYING
+
+HOST_BUILD_DEPENDS:=ninja/host
```
Comment (neheb) [inline devel/meson/Makefile]: not needed. ninja already depends on its host build.

**Comment by neheb** [comment]
Actually...why does the target build depend on the host build?

File: devel/meson/Makefile (hunk 2 above)
Comment (dhewg) [reply devel/meson/Makefile]: Meson requires ninja at runtime, and ports built with meson use `PKG_BUILD_DEPENDS:=meson/host` which then drags ninja/host in with this. How else do I express this (The 'BUILD' part of the name is confusing, but I didn't find anything a la HOSTPKG_DEPENDS)?

Comment (neheb) [reply devel/meson/Makefile]: Right, but ninja is listed as a dependency. ninja/host is picked up from there.

**Comment by dhewg** [comment]
> Actually...why does the target build depend on the host build?

Ninja is built using itself. It has a `--bootstrap` option to build a minimal version of itself which then builds the final binary. This is what the host built does. The target build then uses the host binary to cross compile for the target without the bootstrap option.

File: devel/meson/Makefile (hunk 2 above)
Comment (dhewg) [reply devel/meson/Makefile]: The package does, but that didn't make the host build depend on it IIRC. I can check again though

**Comment by dhewg** [comment]
And that's actually what's happens on my box:
```
Performing configuration checks

    - default address-model    : 32-bit
    - default architecture     : arm

Component configuration:

    - atomic                   : not building
    - chrono                   : not building
    - container                : not building
    - context                  : not building
    - contract                 : not building
    - coroutine                : not building
    - date_time                : not building
    - exception                : building
    - fiber                    : not building
    - filesystem               : not building
    - graph                    : not building
    - graph_parallel           : not building
    - headers                  : building
    - iostreams                : not building
    - locale                   : not building
    - log                      : not building
    - math                     : not building
    - mpi                      : not building
    - program_options          : not building
    - python                   : not building
    - random                   : not building
    - regex                    : not building
    - serialization            : not building
    - stacktrace               : not building
    - system                   : not building
    - test                     : not building
    - thread                   : not building
    - timer                    : not building
    - type_erasure             : not building
    - wave                     : not building
```

With just one file being compiled:
```
gcc.compile.c++ bin.v2/libs/exception/build/gcc-8.3.0/release/link-static/threading-multi/visibility-hidden/clone_current_exception_non_intrusive.o
gcc.archive bin.v2/libs/exception/build/gcc-8.3.0/release/link-static/threading-multi/visibility-hidden/libboost_exception.a
```

But for whatever reason it looks like this on the CI:
```
Performing configuration checks

    - default address-model    : 32-bit
    - default architecture     : mips1
    - C++11 mutex              : yes
    - lockfree boost::atomic_flag : yes
    - Boost.Config Feature Check: cxx11_auto_declarations : yes
    - Boost.Config Feature Check: cxx11_constexpr : yes
    - Boost.Config Feature Check: cxx11_defaulted_functions : yes
    - Boost.Config Feature Check: cxx11_final : yes
    - Boost.Config Feature Check: cxx11_hdr_mutex : yes
    - Boost.Config Feature Check: cxx11_hdr_tuple : yes
    - Boost.Config Feature Check: cxx11_lambdas : yes
    - Boost.Config Feature Check: cxx11_noexcept : yes
    - Boost.Config Feature Check: cxx11_nullptr : yes
    - Boost.Config Feature Check: cxx11_rvalue_references : yes
    - Boost.Config Feature Check: cxx11_template_aliases : yes
    - Boost.Config Feature Check: cxx11_thread_local : yes
    - Boost.Config Feature Check: cxx11_variadic_templates : yes
    - has_icu builds           : no
warning: Graph library does not contain MPI-based parallel components.
note: to enable them, add "using mpi ;" to your user-config.jam.
note: to suppress this message, pass "--without-graph_parallel" to bjam.
    - zlib                     : yes
    - lzma                     : yes
    - zstd                     : yes
    - native-atomic-int32-supported : yes
    - native-syslog-supported  : yes
    - pthread-supports-robust-mutexes : yes
    - gcc visibility           : yes
    - long double support      : yes
    - libbacktrace builds      : no
    - addr2line builds         : no
    - WinDbg builds            : no
    - WinDbgCached builds      : no
    - BOOST_COMP_GNUC >= 4.3.0 : yes
    - zlib                     : yes
    - lzma                     : yes
    - zstd                     : yes

Component configuration:

    - atomic                   : building
    - chrono                   : building
    - container                : building
    - context                  : building
    - contract                 : building
    - coroutine                : building
    - date_time                : building
    - exception                : building
    - fiber                    : building
    - filesystem               : building
    - graph                    : building
    - graph_parallel           : building
    - headers                  : building
    - iostreams                : building
    - locale                   : not building
    - log                      : building
    - math                     : building
    - mpi                      : not building
    - program_options          : building
    - python                   : not building
    - random                   : building
    - regex                    : building
    - serialization            : building
    - stacktrace               : building
    - system                   : building
    - test                     : building
    - thread                   : building
    - timer                    : building
    - type_erasure             : building
    - wave                     : building
```

File: devel/meson/Makefile (hunk 2 above)
Comment (dhewg) [reply devel/meson/Makefile]: If I remove that line, the host meson does not depend on host ninja anymore. So it breaks compiling packages with meson

File: devel/ninja/Makefile [hunk 3]
```diff
@@ -0,0 +1,77 @@
+include $(TOPDIR)/rules.mk
+
+PKG_NAME:=ninja
+PKG_VERSION:=1.9.0
+PKG_RELEASE:=1
+
+PKG_SOURCE:=$(PKG_NAME)-$(PKG_VERSION).tar.gz
+PKG_SOURCE_URL:=https://codeload.github.com/ninja-build/ninja/tar.gz/v$(PKG_VERSION)?
```
Comment (neheb) [inline devel/ninja/Makefile]: codeload should not be used directly. The normal release tarball listed under releases is preferred.

Comment (dhewg) [reply devel/ninja/Makefile]: Ok, I switched meson to that, but the ninja releases are different it seems. The link on the github releases redirects to codeload (`curl "https://github.com/ninja-build/ninja/archive/v1.9.0.tar.gz"|xxd`) and that fails to download, so I kept codeload there.

Comment (neheb) [reply devel/ninja/Makefile]: That seems bizarre. Note that https://github.com/ninja-build/ninja/archive/v1.9.0/ninja-1.9.0.tar.gz works. Well, it redirects to that codeload link. I'd just leave it.

**Comment by neheb** [comment]
Let's just get this in here. No point in delaying.

**Comment by dhewg** [comment]
Nice, thanks!

**Comment by flyn-org** [comment]
Fantastic! This allows us to update a number of packages in OpenWrt that have been awaiting Meson, e.g., https://github.com/openwrt/packages/pull/10433.

Thank you to everyone involved, especially @dhewg and also @br101 and @ja-pa for their earlier work. Discussion and work continued for over a year---I am grateful for the persistence displayed to get this done.

**Comment by dangowrt** [comment]
Somehow CMake is not detected by Meson when trying to build libinput-1.14.3 :(
```
CMake binary for MachineChoice.HOST is not cached
CMake binary missing from cross or native file, or env var undefined.
Found CMake: NO
No CMake binary for machine MachineChoice.HOST not found. Giving up.
```

**Comment by dhewg** [comment]
What do you need cmake for? As far as I understand, it's used as a fallback if a dependency is not found via pkg-config. Does libinput have a dependency without pkg-config support?

**Comment by BKPepe** [comment]
What about to add it to `openwrt-19.07` branch?

**Comment by neheb** [comment]
No opposition from me. That being said, nothing in 19.07 requires it currently.

In other news, I opened up an issue regarding uClibc++: https://github.com/ninja-build/ninja/issues/1684

It's only the hash map feature that's the problem.

**Pull request merged**